Feature/executor example (#39)
jannisspeer/snakemake-executor-plugin-htcondor: examples/job-wrapper/README.md

# Job Wrapper

A simple example demonstrating how Job Wrappers are used with HTCondor executor.

## Why Job Wrappers Are Needed

HTCondor executes jobs in sandboxed environments on remote compute nodes. These nodes don't have access to the user's home directory, which breaks Snakemake's default behavior of writing cache files to `$HOME`.

The job wrapper solves this by:

1. Setting up the proper environment (`HOME=$(pwd)`)
1. Forwarding Snakemake arguments from the access point to the execution point
1. Ensuring all job artifacts stay within the HTCondor scratch directory

### Some common scenarios where job wrapper script is needed:

- The execution point (EP) does not have the right environment to execute and needs modules to be loaded. For example: miniconda (like this example)
- You need to activate a conda environment before anything else runs
- `$HOME` is not set or is pointed to somewhere broken

### Some common scenarios where job wrapper script is not needed:

> - Your workflow uses containers

Reviewer: Is this one completely true? The previous examples all use a container, yet they also use a job wrapper.

Author: I am taking note of this

- You use a shared file-system where the execution point (EP) already have the same environment as the access point (AP)
- EPs already have everything pre-installed

### Illustration

```text
[HTCondor Worker Node]
        │
        ▼
┌─────────────────────────┐
│   job_wrapper.sh        │  ← YOU write this (when needed)
│   - module load conda   │    Sets up the environment
│   - source activate env │
│   - export HOME=$(pwd)  │
│         │               │
│         ▼               │
│  [snakemake_job.sh]     │  ← Snakemake always generates this
│  - run rule `foo`       │    Runs the actual rule
│  - with input X         │
│  - producing output Y   │
└─────────────────────────┘
```

**Notes:**

> Snakemake automatically generates a job script for each rule execution. The job wrapper is not a replacement for this. Instead, it is to ensure that the worker node environment is correctly configured before Snakemake's auto-generated script runs.

Reviewer: I'm afraid it's not true that Snakemake generates a job script for rule execution. Snakemake generates a list of arguments that are passed to the job's executable via condor's `arguments` submit description key. When a job wrapper is used, these arguments need to be passed onto some invocation of Snakemake, which is why most job wrappers end in `snakemake "$@"`

Author: noted

## How This Example Works

This example runs a simple two-step pipeline that processes two samples (`sample1.txt`, `sample2.txt`) through a series of rules:

1. **`make_intermediary`** — Processes each input file, appending `"foo"`
   to produce `results/intermediary_{sample}.txt`
1. **`make_output`** — Processes each intermediary file, appending `"bar"`
   to produce the final `results/output_{sample}.txt`

The pipeline exists purely to demonstrate the job wrapper mechanic — the rules themselves do minimal work so the focus stays on how `wrapper.sh` sets up the environment before each rule executes on the worker node.

Refer to `wrapper.sh` to see how a simple wrapper script is set up.

### Quick Start

This example requires a portable conda environment. Create it with these 3 steps:

**Step 1: Create and pack the environment**

```bash
conda install -c conda-forge conda-pack
conda create -n snakemake-workflow-env -c conda-forge python
conda activate snakemake-workflow-env
pip install snakemake snakemake-executor-plugin-htcondor

> cd /path/to/this/example/job-wrapper

Reviewer: There's a missing `conda deactivate` step here -- conda-pack gets installed into whatever environment was activated before activating `snakemake-workflow-env`, so the command won't exist in `snakemake-workflow-env`.

I'd also recommend changing these to `mamba` commands:
```
mamba install -c conda-forge conda-pack
mamba create -n snakemake-workflow-env -c conda-forge python
mamba activate snakemake-workflow-env
pip install snakemake snakemake-executor-plugin-htcondor
cd /path/to/this/example/job-wrapper
mamba deactivate
conda pack -n snakemake-workflow-env
```

You should also update conda --> mamba in elsewhere, e.g. in the wrapper. Mamba is a lot faster than conda.

> conda pack -n snakemake-workflow-env

Reviewer: I think this should be `conda-pack`, not `conda pack`.

Author: They both are valid commands in this case. CHTC also uses `conda pack`, but I will just change it to `conda-pack`.

#### Prerequisites

1. Install `conda-pack` on your AP:

```bash
conda install -c conda-forge conda-pack
```

2. Create a Clean `conda` Environment

```bash
# environment name: snakemake-workflow-env
conda create -n snakemake-workflow-env -c conda-forge python
# activate the environment
conda activate snakemake-workflow-env
# install executor and Snakemake
pip install snakemake snakemake-executor-plugin-htcondor
```

3. Pack the Environment

```bash
# step 1: cd into the directory where you will be running your workflow
# step 2: pack the environment there. This creates snakemake-workflow-env.tar.gz (~191MB compressed)
conda pack -n snakemake-workflow-env
```

4. Verify the File

```bash
ls -sh snakemake-workflow-env.tar.gz
chmod 644 snakemake-workflow-env.tar.gz
```

5. Update HTCondor Profile
   In `htcondor_profile/config.yaml`, add these under `default-resources`:

```yaml
job_wrapper: "wrapper.sh"
# relative path to submit directory
htcondor_transfer_input_files: "snakemake-workflow-env.tar.gz"
request_disk: "4GB"
request_memory: "1GB"
```

6. Update the wrapper script to include the environment. Refer to `wrapper.sh` for detailed explanations.

> **Note:** The `--dest-prefix` parameter is optional. Without it, the environment unpacks with `bin/`, `lib/`, etc. at the top level, which is simpler for job wrapper scripts.

Reviewer: I still think this prerequisites section should be deleted because it's more or less the same as what you provide in a few commands earlier, but I don't see this `--dest-prefix` flag mentioned anywhere else.

> Refer to the [official CHTC website for more details on creating portable Python installations with Miniconda](https://chtc.cs.wisc.edu/uw-research-computing/conda-installation#4-check-size-of-conda-environment-tar-archive)

Reviewer: Can you move this to the section under where you provide all the conda-pack commands?

# Job Wrapper

A simple example demonstrating how Job Wrappers are used with HTCondor executor.

## Why Job Wrappers Are Needed

HTCondor executes jobs in sandboxed environments on remote compute nodes. These nodes don't have access to the user's home directory, which breaks Snakemake's default behavior of writing cache files to `$HOME`.

The job wrapper solves this by:

1. Setting up the proper environment (`HOME=$(pwd)`)
1. Forwarding Snakemake arguments from the access point to the execution point
1. Ensuring all job artifacts stay within the HTCondor scratch directory

### Some common scenarios where job wrapper script is needed:

> - Compute nodes (EP) do not have the right environment to execute and need modules to be loaded. For e.g: miniconda

Reviewer 2: ```suggestion
- The execution point (EP) does not have the right environment to execute and needs modules to be loaded.
```

Conda example is described in the next item.

- You need to activate a conda environment before anything else runs
- `$HOME` is not set or is pointed to somewhere broken

### Some common scenarios where job wrapper script is not needed:

- Your workflow uses containers

> - You use a shared-file-system where compute nodes (EP) have the same environment as the access point (AP)

Reviewer 2: ```suggestion
- You use a shared file-system where the execution point (EP) has the same environment as the access point (AP)
```
